apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/internal/AJ.kt

```diff
@@ -0,0 +1,45 @@
+package com.apollographql.apollo.cache.normalized.internal
+
+import com.apollographql.apollo.api.ResponseField
+
+/**
+ * A hierarchy of Json object with their fields attached used for normalization
+ */
+sealed class AJElement
+
+/**
+ * field might be null for the root object
+ */
+class AJObject(val fields: List<Field>) : AJElement() {
+  data class Field(val fieldKey: String, val field: ResponseField, val element: AJElement)
+}
+class AJList(val elements: List<AJElement>) : AJElement()
+
+class AJScalar(val value: Any?) : AJElement()
+
+fun AJElement.mergeWith(other: AJElement): AJElement {
+  return when {
+    this is AJScalar && other is AJScalar -> {
+      check(this.value == other.value)
+      this
+    }
+    this is AJList && other is AJList -> throw UnsupportedOperationException()
```

Comment:
hm why is that? We can't merge lists?

Reply:
Not sure why I did that, I guess we need to merge lists nested in arrays indeed. I just added it.